GEODE-7083: Adds statistics to PeerTypeRegistration

pivotal-jbarrett opened the pull request:
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

pivotal-jbarrett commented:
I defaulted to the interface and class name but don’t have a strong preference either way. PDX is likely more recognizable by customers.

demery-pivotal commented:
Do we want meters in addition to (or instead of) statistics?

demery-pivotal commented on geode-core/src/main/java/org/apache/geode/pdx/internal/TypeRegistrationStatistics.java:
```diff
@@ -0,0 +1,56 @@
+import org.apache.geode.Statistics;
+import org.apache.geode.StatisticsFactory;
+import org.apache.geode.StatisticsType;
+
```
With a new class, we have an opportunity to name the instrumentation methods to focus on reporting stuff, rather than on manipulating statistics. For example: Something like `pdxTypeDefined()` and `pdxTypeCreated()`.

Similar for the class name… Is there a good name that focuses on reporting occurrences of interesting events, rather than on statistics? I don't have a great suggestion for what to name the class. My not-necessarily-great attempt is `TypeRegistrationReporter`.

pivotal-jbarrett replied:
I kind of like this idea but what other sort of stuff would we put in these methods? It wouldn't make sense for them to write to the log. So outside of stats, what are you thinking?

If we limit it to stats I still think changing the method names is a reasonable ask. It certainly makes it more readable.

pivotal-jbarrett replied:
I left the class name the same but did change the methods to be more readable. Thoughts?

demery-pivotal replied:
My hope for the future is this: We can hide a certain amount of logging and measuring in the same instrumentation class. In my future fantasy world, the code reports an event to the instrumentation class, and the instrumentation class logs, measures, or both.

A suffix more general than 'stats' might invite us to put some logging in there. The best I can think of is 'Instrumentation' (as in some of the examples from [Pete Hodgson's instrumentation article](https://martinfowler.com/articles/domain-oriented-observability.html)).

I'd like to move in that direction at some point. I recognize that shifting the style will put an extra burden on whoever writes the first few new instrumentation classes.

So given all of that (non-blocking) input, do what you think best.

pivotal-jbarrett replied:
I think out of scope for this because we really need a proposal and examples to drive this. I don't want to trailblaze this.

pivotal-jbarrett commented:
Only CI failure is known issue [GEODE-6499](https://issues.apache.org/jira/browse/GEODE-6499).

pivotal-jbarrett commented:
@upthewaterspout changed the stats name and added enum. Also added a new distributed test to verify the behavior of the define/create/size across servers.

gesterzhou commented on geode-core/src/main/java/org/apache/geode/pdx/internal/PeerTypeRegistration.java:
```diff
@@ -107,18 +110,21 @@ public PeerTypeRegistration(InternalCache cache) {
-    this.dsId = distributedSystemId << 24;
-    this.maxTypeId = 0xFFFFFF;
+    this.distributedSystemId = distributedSystemId << 24;
+    maxTypeId = 0xFFFFFF;
```
it's better to be this.maxTypeId = 0xFFFFFF;

pivotal-jbarrett replied:
`this` is redundant in that context.

pivotal-jbarrett replied:
Did some readability cleanup on the class, so the members being set from parameters are more clear.
